src/Single.php

```diff
@@ -229,19 +229,41 @@ public static function flatten(iterable $data, int $dimensions = 1): \Generator
     /**
      * Group data by a common data element.
      *
-     * The groupKeyFunction determines the key to group elements by.
+     * The groupKeyFunction determines the key (or multiple keys) to group elements by.
+     *
+     * The itemKeyFunction (optional) determines the key of element in group.
      *
      * @param iterable<mixed> $data
      * @param callable        $groupKeyFunction
+     * @param callable|null   $itemKeyFunction
      *
      * @return \Generator<mixed>
      */
-    public static function groupBy(iterable $data, callable $groupKeyFunction): \Generator
-    {
+    public static function groupBy(
+        iterable $data,
+        callable $groupKeyFunction,
+        callable $itemKeyFunction = null
+    ): \Generator {
         $groups = [];
-        foreach ($data as $groupItem) {
-            $groups[$groupKeyFunction($groupItem)][] = $groupItem;
+
+        foreach ($data as $item) {
+            $group = $groupKeyFunction($item);
+            $itemKey = $itemKeyFunction !== null
+                ? $itemKeyFunction($item)
+                : null;
+            $itemGroups = is_iterable($group)
+                ? Transform::toArray($group)
```

Comment:
Question: What is the purpose of this transform to array here? `$itemGroups` is only going to be used in a `foreach` loop, so any iterable type would do here, correct? So, could this be simply?
```php
$itemGroups = \is_iterable($group) ? $group : [$group]
```

Reply:
Yes, you are true! It's my mistake.